# Importing Data

In [1]:
# importing libraries
import pandas as pd
import numpy as mp
import os

In [2]:
# assigning path
path=r'C:\Users\Gal-E\CF Data Analysis\Achievement 4\03-2024 Instacart Basket Analysis'

# import products dataframe from "Original Data" folder
df_prods=pd.read_csv(os.path.join(path,'02 Data','Original Data','products.csv'),index_col=False)

# import orders_wrangled dataframe from "Prepared Data" folder
df_ords=pd.read_csv(os.path.join(path,'02 Data','Prepared Data','orders_wrangled.csv'),index_col=False)

In [3]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [4]:
# the first column is not needed it is the row number from previous dataframe before i exported it 
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [5]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [32]:
df_prods.describe().apply(lambda x: x.apply('{0:.5f}'.format))

product_id     aisle_id department_id       prices
count  49693.00000  49693.00000   49693.00000  49693.00000
mean   24844.34514     67.77025      11.72843      9.99414
std    14343.71740     38.31677       5.85028    453.51969
min        1.00000      1.00000       1.00000      1.00000
25%    12423.00000     35.00000       7.00000      4.10000
50%    24845.00000     69.00000      13.00000      7.10000
75%    37265.00000    100.00000      17.00000     11.20000
max    49688.00000    134.00000      21.00000  99999.00000

In [34]:
df_prods['prices'].value_counts(dropna = False)

prices
2.5        470
5.3        458
6.2        451
2.6        447
5.4        444
          ... 
21.0         1
99999.0      1
21.4         1
15.6         1
18.3         1
Name: count, Length: 242, dtype: int64

In [44]:
# replacing the outliers with median of 'prices'
median = df_prods.loc[df_prods['prices']<25, 'prices'].median()
df_prods['prices'] =  df_prods['prices'].mask(df_prods['prices']>25, median)

In [45]:
df_prods['prices'].value_counts(dropna = False)

prices
2.5     470
5.3     458
6.2     451
2.6     447
5.4     444
       ... 
15.8      2
21.0      1
21.4      1
15.6      1
18.3      1
Name: count, Length: 240, dtype: int64

# Mixed-Type Data

In [6]:
# Create a dataframe to test
df_test = pd.DataFrame()

In [7]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# checking wheter a datafram contains any mixed-type columns
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
# changing data type 
df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

In [46]:
# counting missing values from Products.csv isnull() finds missing values in for of True or False.
# and sum() function obtain the total number of missing observations. true = 1, false = 0
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [47]:
# getting that 16 values from the "df_prods" by creating new dataframe using isnull()
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [48]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [49]:
df_prods.shape

(49693, 5)

In [50]:
# creating new dataframe without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [51]:
df_prods_clean.shape

(49677, 5)

# Duplicates

In [52]:
# looking for full duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [53]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [54]:
df_prods_clean.shape

(49677, 5)

In [55]:
# creating new dataframe that doesnt include duplicate
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [56]:
df_prods_clean_no_dups.shape

(49672, 5)

In [57]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      7.680414
std    14340.705287     38.315784       5.850779      4.199298
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

In [61]:
# Check for mixed-type data in your df_prods dataframe.
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].map(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)
else :
    print ('no colums with mixed data found')

product_name
no colums with mixed data found


# Exporting 'df_prods'

In [68]:
# Export the df_prods_clean_no_dups dataframe as “products_checked.csv” in “Prepared Data” folder.
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# TASK PROCEDURES 4.5 df_ords dataframe

## 2.

In [58]:
# Run the df.describe() function on your df_ords dataframe
# share in a markdown cell whether anything about the data looks off or should be investigated further
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [59]:
# running descriptive statistics for each numeric column and suppress scientific notation

df_ords.describe().apply(lambda x: x.apply('{0:.5f}'.format))

order_id        user_id   order_number orders_day_of_week  \
count  3421083.00000  3421083.00000  3421083.00000      3421083.00000   
mean   1710542.00000   102978.20806       17.15486            2.77622   
std     987581.73982    59533.71779       17.73316            2.04683   
min          1.00000        1.00000        1.00000            0.00000   
25%     855271.50000    51394.00000        5.00000            1.00000   
50%    1710542.00000   102689.00000       11.00000            3.00000   
75%    2565812.50000   154385.00000       23.00000            5.00000   
max    3421083.00000   206209.00000      100.00000            6.00000   

      order_hour_of_day days_since_prior_order  
count     3421083.00000          3214874.00000  
mean           13.45202               11.11484  
std             4.22609                9.20674  
min             0.00000                0.00000  
25%            10.00000                4.00000  
50%            13.00000                7.00000  
75%            16.00000               15.00000  
max            23.00000               30.00000

### havent found anything suspicious. except some missing values in 'days_since_prior_order' column and in 'order_number' column the max value is 100

## 3.

In [60]:
# Check for mixed-type data in your df_ords dataframe.
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
else :
    print ('no colums with mixed data found')

no colums with mixed data found


## 4.

### there were no mixed-type data to fix

## 5.

In [62]:
# checking for missing values in df_ords dataframe
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### in a 'days_since_prior_order' column we found 206209 missing values. to decide what we could to with these missing values we should investigate further

## 6.

In [63]:
# to view the missing values, creating a subset of the dataframe containing only those values
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [64]:
df_nan_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

### counts of the missing values matches the values in 'user_id' column. we have 206209 users. and it is 'order_number' is always '1'
### order_number = 1 indicates the first order of a user
### i would leave these values as they are.

## 7.

In [65]:
# checking duplicate values in df_ords dataframe
df_ords_dups = df_ords[df_ords.duplicated()]

In [66]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

### No duplicates were found

## Exporting 'df_ords'

In [67]:
# Export the df_ords dataframe as “orders_checked.csv” in “Prepared Data” folder.
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))